### PR TITLE
initial proposal for update to atmospheric observation r2d2 ingest

### PR DESCRIPTION
**Pull request by BenjaminRuston**
## Description

this updates the way the atmospheric observations are stored into the r2d2 database using a single python script and then a series of yaml for the observation types

this address issues #11

### CODE REVIEW

File: scripts/obs/r2d2_populate.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    window_length = config.window_length
+    source_path = config.source_path
+    obs_type = config.obs_type
+
+    dates = date_sequence(config.start, config.end, config.window_length)
+    for date in dates:
+        window_start = JediDate(date)
+        for atype in obs_type:
+            if 'windborne' in atype:
+                source_file = f"{source_path}/{window_start.year()}-{window_start.month()}-{window_start.day()}T{window_start.hour()}:{window_start.minute_int():02d}:{window_start.second_int():02d}Z/{provider}.{type}.{atype}.{window_start.year()}-{window_start.month()}-{window_start.day()}T{window_start.hour()}:{window_start.minute_int():02d}:{window_start.second_int():02d}Z.{window_length}.nc4"
```
Comment (ericlingerfelt) [inline scripts/obs/r2d2_populate.py:30]: I recommend that all 'f-strings' use single quotes.

File: scripts/obs/r2d2_populate.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            elif 'ldm' in atype:
+                source_file = f"{source_path}/{window_start.year()}-{window_start.month()}-{window_start.day()}T{window_start.hour()}:{window_start.minute_int():02d}:{window_start.second_int():02d}Z/{provider}.{type}.{atype}.{window_start.year()}-{window_start.month()}-{window_start.day()}T{window_start.hour()}:{window_start.minute_int():02d}:{window_start.second_int():02d}Z.{window_length}.nc4"
+            else:
+                source_file = f"{source_path}/{window_start.year()}{window_start.month()}{window_start.day()}T{window_start.hour()}Z_{window_length}_{atype}.nc4"
+            store(
```
Comment (ericlingerfelt) [inline scripts/obs/r2d2_populate.py:35]: This is using the old R2D2 API. This store call should use ...

File: scripts/obs/r2d2_populate.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#
+# (C) Copyright 2020-2023 UCAR
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+
+import sys
+
+from r2d2 import store
```
Comment (ericlingerfelt) [inline scripts/obs/r2d2_populate.py:12]: `from r2d2 import R2D2Data`

File: scripts/obs/r2d2_populate.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+from solo.date import date_sequence, JediDate
+
+
+def store_observations(yaml_file):
+    config = Configuration(yaml_file)
+    provider = config.provider
+    type = config.type
+    window_length = config.window_length
+    source_path = config.source_path
+    obs_type = config.obs_type
```
Comment (ericlingerfelt) [inline scripts/obs/r2d2_populate.py:23]: There is no `obs_type` in R2D2 V2. We use `observation_type` now.

File: scripts/obs/r2d2_populate.py (hunk 2 above)
Comment (ericlingerfelt) [inline scripts/obs/r2d2_populate.py:35]: Also, `R2D2Data.store` returns the new observation index with each call. We should add a list here to collect these indices. That is, 

```
obs_index_list = []
obs_index = R2D2Data.store()
obs_index_list.append(obs_index)
return obs_index_list
```

**Comment by jeromebarre** [comment]
We discussed this with @BenjaminRuston recently and we think the directory structure in those scripts needs reformatting and avoid duplication of the same or very similar python storage script. This PR is a first step towards this. Ideally we should have one python script that takes a yaml per data stream. Should we organise a mini code sprint during AOP23 to tackle this? More than happy to help!

**Comment by ericlingerfelt** [comment]
Hi @BenjaminRuston, I left you guys some comments. My apologies for the late documentation for R2D2 V2 functions. We should be finishing up the new documentation by the end of next week.

**Comment by BenjaminRuston** [comment]
@ericlingerfelt thanks!  I just hucked this up as this is what we had so yes it needs updating to the new format

hopefully we can discuss next week about the directory reorganization but this seems to follow more what is in `soca/obs` but hopefully we can have a pow-wow and figure something out going forward

**Comment by ashley314** [comment]
Hi, I have a couple of questions to help get a clearer picture of the process. Are all of the source files going to have .nc4 as the extension? Are you going to manually call the new [r2d2_populate.py](https://github.com/JCSDA-internal/r2d2-ingest/pull/12/files/d5c03c7809b0a612a0aa48b7a281d10925863fdb#diff-c8d5c2e2c763727bbb668153883a126bb77d95cae8876e81bc39478ca0d08196) script for each obs/yaml? Who or what process updates those yaml files? I think is is a great start to get files to R2D2! I also was thinking about the idea of using yaml configuration with python scripts to automate some of your pulls. Once we get this in place and working it would be a great tool in moving towards that real-time system.

**Comment by BenjaminRuston** [comment]
the bigger picture for me is there are `ewok` scripts which are called by a meta schedular be it `cylc` or `ecflow` which do the ingest, conversion and then this final step of putting the files into the r2d2 database

**Comment by jeromebarre** [comment]
@ashley314 @BenjaminRuston  this is what we use in COMPO for ingest+conversion
https://github.com/jeromebarre/get_obs
this is far from being complete and needs improvements and tests, but it works for what we need to do in COMPO. Having python scripts in ewok is the way forward!
The difficulty lie in the in the **acquisition** from outside (I wouldn't call this ingest as this can be somewhat misleading with what other call ingest) from obs providers that can differ greatly per product. 
Some provide API calls some just wget or ftp, or you even have the horrible noaa class system... data granularity of the product vary, so you want to make sure you gather or slice the data correctly to fit in our currently 6h data files. This will change in the future and we should store at least hourly files... We need to discuss a strategy that would work for everyone and not just a NWP NRT system!

**Comment by ytremolet** [comment]
> Hi, I have a couple of questions to help get a clearer picture of the process. Are all of the source files going to have .nc4 as the extension? Are you going to manually call the new [r2d2_populate.py](https://github.com/JCSDA-internal/r2d2-ingest/pull/12/files/d5c03c7809b0a612a0aa48b7a281d10925863fdb#diff-c8d5c2e2c763727bbb668153883a126bb77d95cae8876e81bc39478ca0d08196) script for each obs/yaml? Who or what process updates those yaml files? I think is is a great start to get files to R2D2! I also was thinking about the idea of using yaml configuration with python scripts to automate some of your pulls. Once we get this in place and working it would be a great tool in moving towards that real-time system. 

@ashley314 Now that you are getting a better picture of what is there, I think it would be nice to have a chat about where we want to go long term.

File: scripts/obs/r2d2_populate.py (hunk 4 above)
Comment (BenjaminRuston) [reply scripts/obs/r2d2_populate.py:16]: yes the `r2d2_populate.py` script uses `observation_type` in line 30

this `obs_type` is a list generally over different platforms for the data such as:
`atms_npp`, `atms_n20` and `atms_n21`
or
`radiosonde_prepbufr`, `radiosonde_tac` and `radiosonde_ldm`

**Comment by BenjaminRuston** [comment]
@ashley314 should we discuss if we want to use this or incorporate

this is what is being used now for the OBS team to explore new sensors not currently in archive and to populate local r2d2 if needed

File: scripts/obs/r2d2_list.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        adate=f'{window_start.year()}{window_start.month()}{window_start.day()}T{window_start.hour()}0000Z'
+        search_list = R2D2Data.search(item='observation', window_length='PT6H', window_start=adate)
+
+        print(f'  observation list for: {adate}')
+        for obs in search_list:
+            obs_type_index = obs['observation_type_index']
+            provider_index = obs['provider_index']
+            r2d2_index = R2D2Index.get_by_index('observation_type', obs_type_index) 
+            r2d2_provider_index = R2D2Index.get_by_index('provider', provider_index) 
+            print(r2d2_provider_index.get_property('name'), r2d2_index.get_property('name'))
```
Comment (BenjaminRuston) [inline scripts/obs/r2d2_list.py:26]: @ericlingerfelt this is not listing the `provider` and the `observation_type` names

how can we add the `platform` to know if the data are accessible

**Comment by ashley314** [comment]
@BenjaminRuston I think these could be useful. They are using similar patterns and structure we are when working on the ingest suite. If the Obs team uses them then I think we should keep them. I can use it as a guide when we add more sensors. Those are my quick thoughts but open to more of a discussion if you want!

**Comment by ericlingerfelt** [comment]
Any update on this PR? Thanks!

**Comment by BenjaminRuston** [comment]
@ashley314 do you have a start on how you will populate r2d2

wonder if these should stick around until that's settled

**Comment by ashley314** [comment]
@BenjaminRuston Thanks for checking in, I am already populating r2d2 with the ewok ingest suite in [storeObservationsRun.py](https://github.com/JCSDA-internal/ewok/blob/develop/src/runtime/storeObservationsRun.py). Take a look, we are using the expid as provider after discussing with Eric.

**Comment by ericlingerfelt** [comment]
This issue has been open since March 30 and last updated on July 17. Is this issue still in progress?

**Comment by fabiolrdiniz** [comment]
Per previous comments here, should we close this PR?

**Comment by BenjaminRuston** [comment]
agree Fabio this likely can now be closed

@ashley314 we actually maintain these in another repository 

https://github.com/BenjaminRuston/jedi_utils/tree/main/src/python

Please re-open if you think you need it but likely not needed anymore.

**Pull request closed without merging**